### PR TITLE
WL-1459, OSPR-2164 | Wysiwyg editor for prompt

### PR DESCRIPTION
**Pull request by hasnain-naveed**
This is a continuation of Open Source PR by @dmitry-viskov that originated here:
https://github.com/edx/edx-ora2/pull/1058

### CODE REVIEW

File: openassessment/xblock/openassessmentblock.py [hunk 1]
```diff
@@ -176,6 +176,12 @@ class OpenAssessmentBlock(MessageMixin,
         help="The prompts to display to a student."
     )
 
+    prompts_type = String(
+        default='text',
+        scope=Scope.content,
+        help="The type of prompt (possible options: 'text', 'html')."
```
Comment (bill-filler) [inline openassessment/xblock/openassessmentblock.py]: should be translateable

File: openassessment/xblock/test/test_submission.py [hunk 2]
```diff
@@ -124,10 +124,22 @@ def test_closed_submissions(self, xblock):
...
 
+    @scenario('data/prompt_html.xml')
+    def test_prompt_html_to_text(self, xblock):
+        resp = self.request(xblock, 'render_submission', json.dumps(dict()))
+        expected_prompt = u"<code><strong>Question 123</strong></code>"
+        self.assertIn(expected_prompt, resp)
+
+        xblock.prompts_type = "text"
+        resp = self.request(xblock, 'render_submission', json.dumps(dict()))
+        expected_prompt = "&lt;code&gt;&lt;strong&gt;Question 123&lt;/strong&gt;&lt;/code&gt;"
```
Comment (hasnain-naveed) [inline openassessment/xblock/test/test_submission.py]: @bill-filler here is the test case with escaped tags?

**Comment by bill-filler** [comment]
Something is broken when you try to remove a prompt. Open inspector window and you'll see the error:
Uncaught TypeError: Cannot read property 'focus' of undefined
    at OpenAssessment.Prompt.attachWysiwygToPrompt (oa_container_item.js:112)
    at new OpenAssessment.Prompt (oa_container_item.js:80)
    at OpenAssessment.Container.createContainerItem (oa_container.js:75)
    at HTMLDivElement.<anonymous> (oa_container.js:96)
    at HTMLDivElement.dispatch (jquery.js:4737)
    at HTMLDivElement.elemData.handle (jquery.js:4549)
attachWysiwygToPrompt @ oa_container_item.js:112
OpenAssessment.Prompt @ oa_container_item.js:80
createContainerItem @ oa_container.js:75
(anonymous) @ oa_container.js:96
dispatch @ jquery.js:4737
elemData.handle @ jquery.js:4549

File: openassessment/xblock/static/js/src/studio/oa_container_item.js [hunk 3]
```diff
@@ -76,11 +76,44 @@ OpenAssessment.ItemUtilities = {
...
+            var textarea = $(el).find('textarea');
+            var text = $(textarea).val();
+            var type = $(textarea).data('type');
+            if (text && (type === 'text')) {
+                text = _.escape(text).replace(/(?:\r\n|\r|\n)/g, '<br />');
+                $(textarea).val(text);
+            }
+            var newElId = Date.now() + '-textarea-' + (Math.random() * 100);
+            $(textarea).attr('id', newElId).tinymce(tinymceCfg());
+            window.tinyMCE.get(newElId).focus();
```
Comment (bill-filler) [inline openassessment/xblock/static/js/src/studio/oa_container_item.js]: this line is causing the error on remove of prompt. Works fine when creating the prompt. Probably just need to add a check that the element is not null before calling focus on it

Comment (bill-filler) [reply openassessment/xblock/static/js/src/studio/oa_container_item.js]: you should probably add a test for removing the prompt as well as adding one

**Comment by srpearce** [comment]
@bill-filler Can you confirm that this updated documentation is still correct for this feature? Thanks.

https://github.com/edx/edx-documentation/pull/1633

**Comment by bill-filler** [comment]
@hasnain-naveed 
A few issues:
1. The new JS test is failing
2. Inserting an image into the prompt is not working. Do the following to test:
- In studio go to Files + Uploads and upload an image
- copy the link by pressing the Studio button in Files + Uploads
- In the prompt, press "Insert Image" button and paste the link in the Source field
- set a default size 640x480
Upon save, the image is not displayed and it should be. 
If you do the exact same thing by creating an HTML problem in the course you can correctly insert an image using the same image and process.
I'm guessing we don't have the path configured properly when setting up the html editor in the ora prompt - seems like a path issue.

**Comment by bill-filler** [comment]
@hasnain-naveed 
Also make sure we can properly insert a video as well as an image, should work the same way.

**Comment by hasnain-naveed** [comment]
@bill-filler I have diagnosed the bug. It is happening due to file/media uploading feature. It is returning the wrong URL. 

For example, when you upload "Course Card Image" (Settings > Schedule and Details) then its image URL looks like this
 `/asset-v1:edx+CS10101+20189+type@asset+block@FB_IMG_1472756513343.jpg` 
which is the correct URL if you want to use in the course but file uploading feature returns the URL like 
`http://localhost:8001/asset-v1:edx+CS10101+20189+type@asset+block@FB_IMG_1472756513343.jpg` (WEB URL)

So if you use the correct URL in prompt then it is working fine. I am attaching the screenshots. 

adding image


studio

**Comment by bill-filler** [comment]
@hasnain-naveed 
I think you've misunderstood the problem. The way to upload images for the course is to goto Content->Files and Uploads. From there, you click on the "Studio" button next to the image you want to import.
That will copy a a link like "/static/images_course_image.jpg" to the clipboard. Then you paste that link into the Source field for the Image editor dialog. We need to make the component work with this type of path.

The exact same thing works if you add a new Unit in studio, with a component of type "HTML->Text". That uses the same HTML editor we are using. Pasting links of the form "/static/images_course_image.jpg" work  correctly in the image editor there. We need to make ours work the same way

**Comment by ziafazal** [comment]
@hasnain-naveed @bill-filler i have made necessary changes in this PR to allow embedding of studio assets in WYSIWYG prompts.

**Comment by edx-webhook** [comment]
@hasnain-naveed 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-ora2&entry.867055334=https://github.com/edx/edx-ora2/pull/1088&entry.1484655318=https://github.com/hasnain-naveed&entry.752974735=2018-03-14+13:25&entry.1917517419=2018-03-27+13:37&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**